chapter_hyperparameter_optimization/hyperopt-intro.md

```diff
@@ -0,0 +1,209 @@
+like batch size or number of epochs are also worth tuning, we focus on learning
+rate alone for simplicity.
+
+```{.python .input}
+%%tab pytorch
+from d2l import torch as d2l
+
+import numpy as np
+import torch
+from torch import nn
+from scipy import stats
+```
+
+```{.python .input  n=9}
+%%tab pytorch
+class SoftmaxClassification(d2l.Classifier): #@save
+    def __init__(self, num_outputs, lr):
+        super().__init__()
+        self.save_hyperparameters()
+        self.net = nn.Sequential(nn.Flatten(),
+                                 nn.LazyLinear(num_outputs))
+    def forward(self, X):
+        return self.net(X)
+```
+
+Before we can run HPO, we first need to define two ingredients: the objective function and the configuration space.
+
+### The Objective Function
+
+
+The performance of a learning algorithm can be seen as a function $f: \mathcal{X} \rightarrow \mathbb{R}$ that maps from the hyperparameter space $\mathbf{x} \in \mathcal{X}$ to the validation loss. For every evaluation of $f(\mathbf{x})$, we have to train and validate our machine learning model, which can be time and compute intensive in the case of deep neural networks trained on large datasets. Now, given our criterion $f(\mathbf{x})$ our goal is to find $\mathbf{x}_{\star} \in argmin_{\mathbf{x} \in \mathcal{X}} f(\mathbf{x})$. 
+
+There is no simple way to compute gradients of $f$ with respect to $\mathbf{x}$, because it would require to propagate the gradient through the entire training process. While there is recent work :cite:`maclaurin-icml15,franceschi-icml17a` to drive HPO by approximate "hypergradients", none of the existing approaches are competitive with the state-of-the-art yet, and we will not discuss them here. Furthermore, the computational burden of evaluating $f$ requires HPO algorithms to approach the global optimum with as few samples as possible.
+
+The training of neural networks is stochastic (e.g., weights are randomly initialized, mini-batches are randomly sampled), so that our observations will be noisy: $y \sim f(\mathbf{x}) + \epsilon$, where we usually assume that the $\epsilon \sim N(0, \sigma)$ observation noise is Gaussian distributed.
+
+Faced with all these challenges, we usually try to identify a small set of well performing hyperparameter configurations quickly, instead of hitting the global optima exactly. However, due to large computational demands of most neural networks models, even this can take days or weeks of compute. We will explore in :numref:`sec_mf_hpo` how we can speed-up the optimization process by either distributing the search or using cheaper-to-evaluate approximations of the objective function.
+
+
+Now, since we would like to optimize the validation error, we need to add a function computing this quantity.
+
+```{.python .input  n=8}
+%%tab pytorch
+
+@d2l.add_to_class(d2l.Trainer)  #@save
+def validate(self):
```

Comment:
Instead of directly modifying the existing Trainer class in the d2l lib, please:
* subclass d2l.Trainer, say HPOTrainer. If you need to use it elsewhere, save it so you have d2l.HPOTrainer
* Implement `validate` in `d2l.HPOTrainer` or call it `validation_step` to be consistent in naming

Reply:
run `d2lbook build lib`: HPOTrainer will be saved to d2l/torch.py